spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java

```diff
@@ -73,40 +67,6 @@ public GcpCloudSqlAutoConfiguration(GcpCloudSqlProperties gcpCloudSqlProperties,
 		this.gcpProperties = gcpProperties;
 	}
 
-	@Bean
-	@ConditionalOnMissingBean(CloudSqlJdbcInfoProvider.class)
-	@Conditional(AppEngineCondition.class)
-	public CloudSqlJdbcInfoProvider appengineCloudSqlJdbcInfoProvider() {
-		CloudSqlJdbcInfoProvider appEngineProvider = new AppEngineCloudSqlJdbcInfoProvider(this.gcpCloudSqlProperties);
-
-		if (LOGGER.isInfoEnabled()) {
-			LOGGER.info("App Engine JdbcUrl provider. Connecting to "
-					+ appEngineProvider.getJdbcUrl() + " with driver "
-					+ appEngineProvider.getJdbcDriverClass());
-		}
-
-		setCredentialsProperty();
-
-		return appEngineProvider;
-	}
-
-	@Bean
-	@ConditionalOnMissingBean(CloudSqlJdbcInfoProvider.class)
-	public CloudSqlJdbcInfoProvider defaultJdbcInfoProvider() {
-		CloudSqlJdbcInfoProvider defaultProvider = new DefaultCloudSqlJdbcInfoProvider(this.gcpCloudSqlProperties);
-
-		if (LOGGER.isInfoEnabled()) {
-			LOGGER.info("Default " + this.gcpCloudSqlProperties.getDatabaseType().name()
-					+ " JdbcUrl provider. Connecting to "
-					+ defaultProvider.getJdbcUrl() + " with driver "
-					+ defaultProvider.getJdbcDriverClass());
-		}
-
-		setCredentialsProperty();
-
-		return defaultProvider;
-	}
-
 	@Bean
 	@Primary
 	public DataSourceProperties cloudSqlDataSourceProperties(DataSourceProperties properties,
```

Comment:
What happens if a `CloudSqlJdbcInfoProvider` could not  be created? I think there is a case missing here. If neither postgres nor mysql socket factory is available, then no instance  should be created and this part will blow up unless I am missing something.

Reply:
That is a corner case.
If there is no MySQL or PostgreSQL drivers on the classpath, then Spring Cloud GCP is not what our user is looking for.
Making debugging for that case a little easier could be useful. However, adding log lines to this method won't help since we won't get to the method itself due to the lack of dependency...

Reply:
What I mean is that, if no `CloudSqlJdbcInfoProvider` could be created this code _must not_ run at all, otherwise you'll have an exception on startup.

Reply:
I agree with @snicoll. `CloudSqlJdbcInfoProvider` should probably be made ` @Autowired(required = false)` and if it's null, just return the passed in `DataSourceProperties`. Or, maybe there is another mechanism to achieve the same thing.

Reply:
Why `@ConditionalOnBean(CloudSqlJdbcInfoProvider.class)` doesn't work here for you?

Reply:
I really like `@ConditionalOnBean`. Will apply.

Reply:
Always thinking in terms of ConditionalOn*Missing*Bean. No one said, it doesn't work. :smile:

Reply:
After trying with `@ConditionalOnBean`, I don't think it does what we need.

From its Javadoc:

```
 * The condition can only match the bean definitions that have been processed by the
 * application context so far and, as such, it is strongly recommended to use this
 * condition on auto-configuration classes only. If a candidate bean may be created by
 * another auto-configuration, make sure that the one using this condition runs after.
```

It looks like the classes in `@Import` are "processed", but that process doesn't generate any beans.
The Javadoc also seems to imply that the beans should be generated by another auto-configuration, and the `@Import` might make it as if they are generated by *this* auto-configuration.

On top of that, the error message is very clear that a certain bean requires another bean of type `CloudSqlJdbcInfoProvider`.
Which is better than when `GcpCloudSqlAutoConfiguration` doesn't run at all:

```
Caused by: org.springframework.boot.autoconfigure.jdbc.DataSourceProperties$DataSourceBeanCreationException: Cannot determine embedded database driver class for database type NONE. If you want an embedded database please put a supported one on the classpath. If you have database settings to be loaded from a particular profile you may need to active it (no profiles are currently active).
```

All things considered, I would like to suggest that if we want to make any changes here, we do so in another issue so as not to drag the fork of the SQL starter.

Reply:
Well, when I add `@AutoConfigureBefore(GcpCloudSqlAutoConfiguration.class)` into the `MySqlJdbcInfoProviderAutoConfiguration` and `PostgreSqlJdbcInfoProviderAutoConfiguration` and I also have that `@ConditionalOnBean(CloudSqlJdbcInfoProvider.class)` on this `cloudSqlDataSourceProperties` all the tests in the `autoconfigure` pass for me.

:shrugging:

Reply:
Interesting... it wasn't working with the `@AutoConfigureBefore` in the other classes and it probably should?
With that change, it's also working for me.

Reply:
That's why auto-configurations should be ordered if they have dependencies. `@ConditionalOnBean` or `@ConditionalOnMissingBean` will not work if the bean definition is processed after the evaluation of the condition (it's not the same thing as an injection point that forces the `ApplicationContext`  to resolve the dependency.